std/algorithm/setops.d

```diff
@@ -387,6 +387,16 @@ auto cartesianProduct(RR...)(RR ranges)
                     r = ranges[i].save; // rollover
             }
         }
+        static if (allSatisfy!(hasLength, RR))
+        @property size_t length()
+        {
+            size_t length = 1;
+            foreach (i, ref r; current)
+            {
+                length *= r.length;
```

Comment:
Maybe an assert? I'm not sure about using `enforce`, it might introduce too much overhead.